File: OpenBurnBarCore/Sources/OpenBurnBarCore/SharedModels/Insights/InsightLayout.swift
import Foundation

/// A deterministic, columnar grid layout for the widgets on a canvas.
///
/// macOS uses 12 columns; iPad projects to 6; iPhone to 2. The
/// projection is a pure function so the same canvas always reflows the
/// same way regardless of which device opens it.
public struct InsightLayout: Codable, Hashable, Sendable {
    /// Intended column count for the authoring platform.
    public var columnCount: Int
    /// Pt height of a single row.
    public var rowHeight: Double
    /// Pt gap between cells.
    public var gap: Double
    /// Widget-id → cell placement.
    public var placements: [UUID: CellPlacement]
    /// Monotonic counter bumped on every mutation; used to merge concurrent
    /// edits from another device (last-revision-wins).
    public var revision: Int

    public init(
        columnCount: Int = 12,
        rowHeight: Double = 96,
        gap: Double = 12,
        placements: [UUID: CellPlacement] = [:],
        revision: Int = 0
    ) {
        self.columnCount = max(1, columnCount)
        self.rowHeight = max(32, rowHeight)
        self.gap = max(0, gap)
        self.placements = placements
        self.revision = revision
    }

    /// Single widget placement.
    public struct CellPlacement: Codable, Hashable, Sendable {
        /// 0-indexed column.
        public var column: Int
        /// 0-indexed row.
        public var row: Int
        /// 1…columnCount.
        public var colSpan: Int
        /// ≥ 1.
        public var rowSpan: Int

        public init(column: Int, row: Int, colSpan: Int, rowSpan: Int) {
            self.column = max(0, column)
            self.row = max(0, row)
            self.colSpan = max(1, colSpan)
            self.rowSpan = max(1, rowSpan)
        }
    }

    /// Insert a new widget at the first free cell large enough to fit
    /// `defaultSpan`. Bumps `revision`.
    public mutating func placeNew(widgetID: UUID, defaultSpan: (columns: Int, rows: Int)) {
        let cols = min(max(1, defaultSpan.columns), columnCount)
        let rows = max(1, defaultSpan.rows)
        let occupancy = makeOccupancyGrid()
        let (c, r) = firstFreeCell(occupancy: occupancy, colSpan: cols, rowSpan: rows)
        placements[widgetID] = CellPlacement(column: c, row: r, colSpan: cols, rowSpan: rows)
        revision &+= 1
    }

    /// Move a widget to a new origin (column, row), clamping the spans so
    /// it stays inside the column count. Bumps `revision`.
    public mutating func move(widgetID: UUID, toColumn column: Int, toRow row: Int) {
        guard var current = placements[widgetID] else { return }
        current.column = max(0, min(column, columnCount - current.colSpan))
        current.row = max(0, row)
        placements[widgetID] = current
        revision &+= 1
    }

    /// Resize a widget by setting new spans. Clamped to fit columnCount.
    /// Bumps `revision`.
    public mutating func resize(widgetID: UUID, colSpan: Int, rowSpan: Int) {
        guard var current = placements[widgetID] else { return }
        let newCol = max(1, min(colSpan, columnCount))
        let newRow = max(1, rowSpan)
        current.colSpan = newCol
        current.rowSpan = newRow
        current.column = min(current.column, max(0, columnCount - newCol))
        placements[widgetID] = current
        revision &+= 1
    }

    /// Remove a widget's placement.
    public mutating func remove(widgetID: UUID) {
        if placements.removeValue(forKey: widgetID) != nil {
            revision &+= 1
        }
    }

    /// Total rows currently occupied (for sizing scroll content).
    public var rowCount: Int {
        placements.values.reduce(0) { max($0, $1.row + $1.rowSpan) }
    }

    /// Project to a different column count. The projection preserves
    /// widget order (row-major) and proportionally clamps spans, never
    /// causing overlaps.
    public func projected(toColumnCount targetCols: Int) -> InsightLayout {
        let target = max(1, targetCols)
        if target == columnCount { return self }

        // Stable order: top-to-bottom, then left-to-right, then by UUID.
        let ordered = placements
            .map { ($0.key, $0.value) }
            .sorted {
                if $0.1.row != $1.1.row { return $0.1.row < $1.1.row }
                if $0.1.column != $1.1.column { return $0.1.column < $1.1.column }
                return $0.0.uuidString < $1.0.uuidString
            }

        var projected: [UUID: CellPlacement] = [:]
        var cursorCol = 0
        var cursorRow = 0
        var rowMaxHeight = 0

        for (id, p) in ordered {
            // Proportional span, rounded, clamped.
            let proportional = Double(p.colSpan) * Double(target) / Double(max(1, columnCount))
            let span = max(1, min(target, Int(proportional.rounded())))

            // Wrap to next row if we can't fit.
            if cursorCol + span > target {
                cursorRow += max(1, rowMaxHeight)
                cursorCol = 0
                rowMaxHeight = 0
            }

            projected[id] = CellPlacement(
                column: cursorCol,
                row: cursorRow,
                colSpan: span,
                rowSpan: p.rowSpan
            )
            cursorCol += span
            rowMaxHeight = max(rowMaxHeight, p.rowSpan)
        }

        return InsightLayout(
            columnCount: target,
            rowHeight: rowHeight,
            gap: gap,
            placements: projected,
            revision: revision
        )
    }

    // MARK: - Private helpers

    /// Build a sparse [row][col] occupancy grid sized to the current spread.
    private func makeOccupancyGrid() -> [[Bool]] {
        let rows = rowCount + 1
        var grid = Array(repeating: Array(repeating: false, count: columnCount), count: max(rows, 1))
        for p in placements.values {
            for r in p.row..<min(p.row + p.rowSpan, grid.count) {
                for c in p.column..<min(p.column + p.colSpan, columnCount) {
                    grid[r][c] = true
                }
            }
        }
        return grid
    }

    /// Find the first row-major free rectangle of size `colSpan × rowSpan`.
    /// If no such rectangle exists within the current grid, append the
    /// widget at row `rowCount` (i.e. directly below the existing content).
    private func firstFreeCell(occupancy: [[Bool]], colSpan: Int, rowSpan: Int) -> (Int, Int) {
        let rows = occupancy.count
        for r in 0..<rows {
            if r + rowSpan > rows { break }
            for c in 0..<columnCount {
                if c + colSpan > columnCount { break }
                if rangeIsFree(occupancy: occupancy, c: c, r: r, colSpan: colSpan, rowSpan: rowSpan) {
                    return (c, r)
                }
            }
        }
        return (0, rowCount)
    }

    private func rangeIsFree(occupancy: [[Bool]], c: Int, r: Int, colSpan: Int, rowSpan: Int) -> Bool {
        for rr in r..<(r + rowSpan) {
            for cc in c..<(c + colSpan) {
                if occupancy[rr][cc] { return false }
            }
        }
        return true
    }
}

File: windows/app/OpenBurnBar.App.Presentation/Insights/InsightLayout.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace OpenBurnBar.App.Presentation.Insights;

/// <summary>
/// A single widget placement in the columnar canvas grid — port of the Swift
/// <c>InsightLayout.CellPlacement</c>. Columns/rows are 0-indexed; spans are clamped ≥ 1 on
/// construction.
/// </summary>
public sealed record CellPlacement
{
    public CellPlacement(int column, int row, int colSpan, int rowSpan)
    {
        Column = Math.Max(0, column);
        Row = Math.Max(0, row);
        ColSpan = Math.Max(1, colSpan);
        RowSpan = Math.Max(1, rowSpan);
    }

    /// <summary>0-indexed column origin.</summary>
    public int Column { get; init; }

    /// <summary>0-indexed row origin.</summary>
    public int Row { get; init; }

    /// <summary>Width in columns (1…columnCount).</summary>
    public int ColSpan { get; init; }

    /// <summary>Height in rows (≥ 1).</summary>
    public int RowSpan { get; init; }
}

/// <summary>
/// A deterministic, columnar grid layout for the widgets on a canvas — a direct port of the
/// macOS <c>InsightLayout</c> (<c>OpenBurnBarCore/.../Insights/InsightLayout.swift</c>).
///
/// macOS uses 12 columns; a narrower host projects to fewer. Both the row-major
/// first-fit packing (<see cref="PlaceNew"/>) and the reflow (<see cref="Projected"/>) are
/// pure functions, so the same canvas always lays out the same way regardless of which host
/// opens it — which is exactly why the whole file is unit-tested on the macOS authoring host.
/// </summary>
public sealed class InsightLayout
{
    public InsightLayout(
        int columnCount = 12,
        double rowHeight = 96,
        double gap = 12,
        IReadOnlyDictionary<Guid, CellPlacement>? placements = null,
        int revision = 0)
    {
        ColumnCount = Math.Max(1, columnCount);
        RowHeight = Math.Max(32, rowHeight);
        Gap = Math.Max(0, gap);
        Placements = placements is null
            ? new Dictionary<Guid, CellPlacement>()
            : new Dictionary<Guid, CellPlacement>(placements);
        Revision = revision;
    }

    /// <summary>Intended column count for the authoring platform.</summary>
    public int ColumnCount { get; }

    /// <summary>Pt height of a single row.</summary>
    public double RowHeight { get; }

    /// <summary>Pt gap between cells.</summary>
    public double Gap { get; }

    /// <summary>Widget-id → cell placement.</summary>
    public Dictionary<Guid, CellPlacement> Placements { get; }

    /// <summary>Monotonic counter bumped on every mutation (last-revision-wins merge).</summary>
    public int Revision { get; private set; }

    /// <summary>Total rows currently occupied (for sizing scroll content).</summary>
    public int RowCount => Placements.Values.Aggregate(0, (acc, p) => Math.Max(acc, p.Row + p.RowSpan));

    /// <summary>
    /// Insert a widget at the first free cell large enough to fit <paramref name="defaultSpan"/>.
    /// Bumps <see cref="Revision"/>.
    /// </summary>
    public void PlaceNew(Guid widgetId, (int Columns, int Rows) defaultSpan)
    {
        int cols = Math.Min(Math.Max(1, defaultSpan.Columns), ColumnCount);
        int rows = Math.Max(1, defaultSpan.Rows);
        bool[][] occupancy = MakeOccupancyGrid();
        (int c, int r) = FirstFreeCell(occupancy, cols, rows);
        Placements[widgetId] = new CellPlacement(c, r, cols, rows);
        Revision++;
    }

    /// <summary>Move a widget origin, clamping the column so its span stays inside the grid.</summary>
    public void Move(Guid widgetId, int toColumn, int toRow)
    {
        if (!Placements.TryGetValue(widgetId, out CellPlacement? current))
        {
            return;
        }

        int column = Math.Max(0, Math.Min(toColumn, ColumnCount - current.ColSpan));
        int row = Math.Max(0, toRow);
        Placements[widgetId] = current with { Column = column, Row = row };
        Revision++;
    }

    /// <summary>Resize a widget's spans, clamped to fit the column count.</summary>
    public void Resize(Guid widgetId, int colSpan, int rowSpan)
    {
        if (!Placements.TryGetValue(widgetId, out CellPlacement? current))
        {
            return;
        }

        int newCol = Math.Max(1, Math.Min(colSpan, ColumnCount));
        int newRow = Math.Max(1, rowSpan);
        int clampedColumn = Math.Min(current.Column, Math.Max(0, ColumnCount - newCol));
        Placements[widgetId] = current with { ColSpan = newCol, RowSpan = newRow, Column = clampedColumn };
        Revision++;
    }

    /// <summary>Remove a widget's placement (bumps revision only if one existed).</summary>
    public void Remove(Guid widgetId)
    {
        if (Placements.Remove(widgetId))
        {
            Revision++;
        }
    }

    /// <summary>
    /// Project to a different column count. Preserves widget order (row-major, then column,
    /// then id) and proportionally clamps spans, never causing overlaps. Direct port of the
    /// Swift <c>projected(toColumnCount:)</c>.
    /// </summary>
    public InsightLayout Projected(int targetColumns)
    {
        int target = Math.Max(1, targetColumns);
        if (target == ColumnCount)
        {
            return this;
        }

        List<KeyValuePair<Guid, CellPlacement>> ordered = Placements
            .OrderBy(kv => kv.Value.Row)
            .ThenBy(kv => kv.Value.Column)
            .ThenBy(kv => kv.Key.ToString(), StringComparer.Ordinal)
            .ToList();

        var projected = new Dictionary<Guid, CellPlacement>();
        int cursorCol = 0;
        int cursorRow = 0;
        int rowMaxHeight = 0;

        foreach ((Guid id, CellPlacement p) in ordered.Select(kv => (kv.Key, kv.Value)))
        {
            double proportional = p.ColSpan * (double)target / Math.Max(1, ColumnCount);
            int span = Math.Max(1, Math.Min(target, (int)Math.Round(proportional, MidpointRounding.AwayFromZero)));

            if (cursorCol + span > target)
            {
                cursorRow += Math.Max(1, rowMaxHeight);
                cursorCol = 0;
                rowMaxHeight = 0;
            }

            projected[id] = new CellPlacement(cursorCol, cursorRow, span, p.RowSpan);
            cursorCol += span;
            rowMaxHeight = Math.Max(rowMaxHeight, p.RowSpan);
        }

        return new InsightLayout(target, RowHeight, Gap, projected, Revision);
    }

    /// <summary>Deep copy (used when stamping a template into a fresh canvas).</summary>
    public InsightLayout Clone() => new(ColumnCount, RowHeight, Gap, Placements, Revision);

    // ── Private helpers ──────────────────────────────────────────────────────────

    private bool[][] MakeOccupancyGrid()
    {
        int rows = RowCount + 1;
        int rowCount = Math.Max(rows, 1);
        var grid = new bool[rowCount][];
        for (int r = 0; r < rowCount; r++)
        {
            grid[r] = new bool[ColumnCount];
        }

        foreach (CellPlacement p in Placements.Values)
        {
            for (int r = p.Row; r < Math.Min(p.Row + p.RowSpan, grid.Length); r++)
            {
                for (int c = p.Column; c < Math.Min(p.Column + p.ColSpan, ColumnCount); c++)
                {
                    grid[r][c] = true;
                }
            }
        }

        return grid;
    }

    private (int Column, int Row) FirstFreeCell(bool[][] occupancy, int colSpan, int rowSpan)
    {
        int rows = occupancy.Length;
        for (int r = 0; r < rows; r++)
        {
            if (r + rowSpan > rows)
            {
                break;
            }

            for (int c = 0; c < ColumnCount; c++)
            {
                if (c + colSpan > ColumnCount)
                {
                    break;
                }

                if (RangeIsFree(occupancy, c, r, colSpan, rowSpan))
                {
                    return (c, r);
                }
            }
        }

        return (0, RowCount);
    }

    private static bool RangeIsFree(bool[][] occupancy, int c, int r, int colSpan, int rowSpan)
    {
        for (int rr = r; rr < r + rowSpan; rr++)
        {
            for (int cc = c; cc < c + colSpan; cc++)
            {
                if (occupancy[rr][cc])
                {
                    return false;
                }
            }
        }

        return true;
    }
}
